sdk/src/azure/core/az_mqtt5_connection_hfsm.c

```diff
@@ -382,7 +464,25 @@ static az_result disconnecting(az_event_policy* me, az_event event)
   switch (event.type)
   {
     case AZ_HFSM_EVENT_ENTRY:
+      _az_RETURN_IF_FAILED(az_platform_timer_create(
+          &this_policy->_internal.connection_timer, _connection_timeout_callback, this_policy));
+      _az_RETURN_IF_FAILED(az_platform_timer_start(
+          &this_policy->_internal.connection_timer,
+          this_policy->_internal.disconnecting_timeout_msec));
+      break;
     case AZ_HFSM_EVENT_EXIT:
+      _az_RETURN_IF_FAILED(az_platform_timer_destroy(&this_policy->_internal.connection_timer));
+      break;
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    {
+      az_mqtt5_connack_data* data = (az_mqtt5_connack_data*)event.data;
+
+      if (data->connack_reason == 0)
+      {
+        _disconnect(this_policy);
```

Comment:
Is this because the previous disconnect didn't go through?

And if the connack_reason isn't 0, should that act like getting the disconnect_rsp in this state?

Reply:
It's due to race/network queuing: the application decided to stop the connection shortly after starting it. The network had a CONNACK in-flight. We receive the CONNACK (and connection completion). It is likely that the server will also honor the DISCONNECT, but there is nothing preventing us from requesting the disconnect again.

Receiving an error CONNACK will switch the state machine to Faulted, which is the simpler solution. Another way would be to run the same code as in ReconnectTimeout (although without actually reconnecting) then switching to Idle. This ensures that credentials get updated prior to a new retry.